src/main/scala/com/wavesplatform/matcher/api/MatcherApiRoute.scala

```diff
@@ -124,12 +127,40 @@ case class MatcherApiRoute(wallet: Wallet,
     (pathEndOrSingleSlash & post) {
       json[Order] { order =>
         placeTimer.measure {
-          (matcher ? order).mapTo[MatcherResponse].map(_.toHttpResponse)
+          (matcher ? order).mapTo[MatcherResponse]
         }
       }
     }
   }
 
+  private def cancelOrder(orderId: ByteStr, senderPublicKey: Option[PublicKeyAccount]): ToResponseMarshallable = {
+    DBUtils.orderInfo(db, orderId).status match {
+      case LimitOrder.NotFound      => StatusCodes.NotFound
+      case status if status.isFinal => StatusCodes.BadRequest -> Json.obj("message" -> s"Order is already ${status.name}")
+      case _ =>
+        DBUtils.order(db, orderId) match {
+          case None =>
+            log.warn(s"Order $orderId was not found in history")
+            StatusCodes.NotFound
+          case Some(order) if senderPublicKey.exists(_ != order.senderPublicKey) =>
+            StatusCodes.BadRequest -> Json.obj("message" -> "Public Key mismatch")
+          case Some(order) =>
+            orderBook(order.assetPair) match {
```

Comment:
What if the matcher is starting (not yet started), and we receive the cancel request?